File: EvaluatePostfix.cpp
// https://leetcode.com/problems/evaluate-reverse-polish-notation/

class Solution {
public:
    bool isOperator(string symbol){
        if (symbol.size()>1){
            return false;
        }
        char opr = symbol[0];
        switch (opr){
            case '+':
            case '-':
            case '*':
            case '/': return true;
        }
        return false;
    }
    
    int getResult(char opr, int first, int second){
        int result = 0;
        switch (opr){
            case '+': result = first + second; break;
            case '-': result = first - second; break;
            case '*': result = first * second; break;
            case '/': result = first / second; break;
        }
        return result;
        
    }
    int evalRPN(vector<string>& tokens) {
        if (tokens.size()==1) return stoi(tokens[0]);
        
        stack<int> s;
        int result = 0;
        
        for (int i=0; i<tokens.size(); i++){
            if (!isOperator(tokens[i])){
                int n = stoi(tokens[i]);
                s.push(n);
            } else {
                int second = s.top();
                s.pop();
                int first = s.top();
                s.pop();
                result = getResult(tokens[i][0], first, second);
                s.push(result);
            }
        }
        
        return result;
    }
};

File: EvaluatePostfix.py
# Evaluate postfix expression

def push(stack, element):
    stack.append(element)

def pop(stack):
    return stack.pop()

def isOperator(symbol):
    return symbol in "+-*/%^"

def evaluatePostfix(postfix):
    stack = []

    postfix = postfix.split(",")

    for symbol in postfix:
        if not isOperator(symbol):
            push(stack, int(symbol))
        else:
            operand2 = pop(stack)
            operand1 = pop(stack)
            result = eval(str(operand1) + symbol + str(operand2))
            push(stack, result)

    return pop(stack)

expressions = ["10,20,+,15,5,+,*", "50,40,+,18,14,-,4,*,+", "10,20,+,25,15,-,*,30,/"]

for expn in expressions:
    print(f"[{expn}] = {evaluatePostfix(expn)}")


'''

[10,20,+,15,5,+,*] = 600
[50,40,+,18,14,-,4,*,+] = 106
[10,20,+,25,15,-,*,30,/] = 10.0

'''
